### PR TITLE
Underwater Image Processing Technique-works well

### PR DESCRIPTION
**Pull request by Si-D**
This is a python version for Enhancement of underwater images with statistical model of background light and optimization of transmission map. 

Dataset for underwater images : https://li-chongyi.github.io/proj_benchmark.html

Here is the list of libraries you need to install to execute the code:

- `python = 3.6`

- `cv2`

- `numpy`

- `scipy`

- `matplotlib`

- `scikit-image`

- `natsort`

**Usage Instructions:**

- Install the necessary packages.

- Create two folders 'InputImages' and 'OutputImages'.

- Put the raw images into the Input Images folder.

- Run python main.py

- You can find the output images in the 'OutputImages' folder.

### CODE REVIEW

**Comment by PratikGarai** [comment]
Please make 2 changes : 
1. Add __pycache__ and any other redundant files to gitignore and remove them from the remote repo
2. Add a readme inside your folder with a mini demonstration of the input and output side-by-side

File: EnhancementofUnderwaterImages/ColorCorrect.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    sceneRadiance = np.zeros((img.shape), 'float64')
+    sceneRadiance[:, :, 0] = b
+    sceneRadiance[:, :, 1] = g
+    sceneRadiance[:, :, 2] = r
+    sceneRadiance = sceneRadiance * 255
+    sceneRadiance = np.clip(sceneRadiance,0, 255)
+    sceneRadiance = np.uint8(sceneRadiance)
+    return sceneRadiance
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/ColorCorrect.py]: Remove these extra lines

File: EnhancementofUnderwaterImages/GuidedFilter.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        return ( 1.0 / 255.0 ) * np.float32(img)
+
+    def _initFilter(self):
+        I = self._I
+        r = self._radius
+        eps = self._epsilon
+
+        Ir, Ig, Ib = I[:, :, 0], I[:, :, 1], I[:, :, 2]
+
+        # self._Ir_mean = cv2.blur(Ir, (r, r))
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/GuidedFilter.py]: explain the reason for commenting these parts and when we might uncomment and use these. Else, remove these lines

File: EnhancementofUnderwaterImages/Saturation.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import natsort
+from skimage.color import rgb2lab, lab2rgb
+import matplotlib.pyplot as plt
+
+
+np.seterr(over='ignore')
+if __name__ == '__main__':
+    pass
+
+path = "F:/PaperExperiments/ACMMM2018/OptimalTM_BLs_Restor/InputImages"
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/Saturation.py]: Do not use fixed paths like these. Files are different in other's computers.
Change these to relative paths.

File: EnhancementofUnderwaterImages/depthMapEstimation.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/depthMapEstimation.py:1]: redundant line

File: EnhancementofUnderwaterImages/depthMapEstimation.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    g = img[:, :, 1]
+    r = img[:, :, 2]
+    x_1 = np.maximum(g, b)
+    x_2 = r
+
+    Deptmap = theta_0 + theta_1 * x_1 + theta_2 * x_2
+
+
+    return Deptmap
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/depthMapEstimation.py]: redundant lines

File: EnhancementofUnderwaterImages/depthMapEstimation.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    theta_2 = -0.90511117
+    img = img / 255.0
+    b = img[:, :, 0]
+    g = img[:, :, 1]
+    r = img[:, :, 2]
+    x_1 = np.maximum(g, b)
+    x_2 = r
+
+    Deptmap = theta_0 + theta_1 * x_1 + theta_2 * x_2
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/depthMapEstimation.py:17]: redundant lines

File: EnhancementofUnderwaterImages/depthMin.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        Max_Abs =  np.absolute(img[i] - BL[i])
+        Max_I = np.max(Max_Abs)
+        Max_B = np.max([BL[i],(1 -BL[i])])
+        temp  = Max_I / Max_B
+        Max.append(temp)
+    K_b = np.max(Max)
+    min_depth = 1 - K_b
+
+    return min_depth
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/depthMin.py]: redundant lines

File: EnhancementofUnderwaterImages/getGBTransmission.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            # if(depth_map[i,j]>15):
+            #     depth_map[i, j] = 15
+            # if (depth_map[i, j] < 1):
+            #     depth_map[i, j] = 1
+
+
+
+    transmissionG = 0.93 ** depth_map
+    transmissionB = 0.95 ** depth_map
+    # transmissionG = np.zeros(transmissionR.shape)
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/getGBTransmission.py:18]: explain why you'd comment out these lines and in which situation the user may uncomment and use them

**Comment by k4u5h1k** [comment]
Also please credit the original repo.  
i.e [Enhancement-of-Underwater-Images-with-Statistical-Model-of-BL-and-Optimization-of-TM](https://github.com/wangyanckxx/Enhancement-of-Underwater-Images-with-Statistical-Model-of-BL-and-Optimization-of-TM)

File: EnhancementofUnderwaterImages/getGBTransmission.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    # print('ratioG',ratioG)
+    # transmissionG = transmissionR ** ratioG
+    # transmissionB = transmissionR ** ratioB
+    # print('getGBTransmissionESttransmissionB',transmissionB)
+    return transmissionB,transmissionG,depth_map
+
+
+
+
+# def getGBTransmissionESt(transmissionR,AtomsphericLightTM):
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/getGBTransmission.py:30]: explanation

File: EnhancementofUnderwaterImages/getRefinedTransmission.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    transmissionB = guided_filter.filter(transmissionB)
+
+    transmission = np.zeros(img.shape)
+    transmission[:, :, 0] = transmissionB
+    transmission[:, :, 1] = transmissionG
+    transmission[:, :, 2] = transmissionR_Stretched
+    transmission = np.clip(transmission,0.05, 0.95)
+
+    return transmission
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/getRefinedTransmission.py:25]: explanation + removal of redundant lines

File: EnhancementofUnderwaterImages/getTransmissionMap.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        for j in range(addSize, newWidth - addSize):
+            localMin = 1
+            for k in range(i - addSize, i + addSize + 1):
+                for l in range(j - addSize, j + addSize + 1):
+                    if imgMiddle.item((k, l)) < localMin:
+                        localMin = imgMiddle.item((k, l))
+            imgDark[i - addSize, j - addSize] = localMin
+        transmission = (1 - imgDark) / (1 - 0.1 / np.max(AtomsphericLight))
+    transmission = np.clip(transmission, 0.1, 0.9)
+    # for i in range(0, transmission.shape[0]):
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/getTransmissionMap.py]: explanation

File: EnhancementofUnderwaterImages/global_histogram_stretching.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import math
+
+
+import numpy as np
+from scipy import stats
+import cv2
+
+pi = math.pi
+e = math.e
+
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/global_histogram_stretching.py:9]: lines...

File: EnhancementofUnderwaterImages/main.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+from global_histogram_stretching import stretching
+from sceneRadiance import sceneRadianceRGB
+
+np.seterr(over='ignore')
+if __name__ == '__main__':
+    pass
+
+starttime = datetime.datetime.now()
+
+folder = "S:/Code Stuff/Image Enhancements/Single-Underwater-Image-Enhancement-and-Color-Restoration/Enhancement-of-Underwater-Images-with-Statistical-Model-of-BL-and-Optimization-of-TM/EnhancementofUnderwaterImages"
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/main.py]: change this to a relative path

File: EnhancementofUnderwaterImages/main_Single.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+    AtomsphericLight = np.zeros(3)
+    AtomsphericLight[0] = (1.13 * np.mean(img[:, :, 0])) + 1.11 * np.std(img[:, :, 0]) - 25.6
+    AtomsphericLight[1] = (1.13 * np.mean(img[:, :, 1])) + 1.11 * np.std(img[:, :, 1]) - 25.6
+    AtomsphericLight[2] = 140 / (1 + 14.4 * np.exp(-0.034 * np.median(img[:, :, 2])))
+    AtomsphericLight = np.clip(AtomsphericLight, 10, 245)
+    print('AtomsphericLight', AtomsphericLight)
+    transmissionR = getTransmission(img, AtomsphericLight, blockSize)
+    transmissionR_new = transmissionR
+    # TM_R_modified = Depth_TM(img, AtomsphericLight)
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/main_Single.py:57]: explaination

File: EnhancementofUnderwaterImages/main_Single.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    # #
+    # #
+    # # cv2.imwrite('Results_TMs/' + Num + 'TM_R.jpg', np.uint8(transmissionR  * 255))
+    # # cv2.imwrite('Results_TMs/' + Num + 'TM_R_Refi_fusion_Sat_lamba.jpg', np.uint8(transmissionR_Stretched  * 255))
+    #
+    #
+    # sceneRadiance = sceneRadianceRGB(img, transmission, AtomsphericLight)
+    # # cv2.imwrite('Results_TMs_minR_0point25/' + Num + 'Restoration_Sat_lamba.jpg', sceneRadiance)
+    # cv2.imwrite('Results_TMs_minR_0point25/' + Num + 'Restoration_Fusion.jpg', sceneRadiance)
+    # #
```
Comment (PratikGarai) [inline EnhancementofUnderwaterImages/main_Single.py]: redundant

**Review by PratikGarai — COMMENTED** [review]
There are parts which I have pointed out... please look into these and try to do the same for subsequent contributions
1. Use relative paths. Your filesystem is different from someone else's. 
2. Remove redundant empty lines from the files.
3. If you are commenting a part of code that involves logic (not something like a simple print statement), then please explain why you commented it out and under which circumstances does the user need to use that block by un-commenting it.
4. Add reference comments/formulae to the code where some complex mathematical operation is involved. If you are not sure, then add references to the original source or some document where someone can read that up.

**Comment by PratikGarai** [comment]
Looks good!

**Comment by PratikGarai** [comment]
@dreadnoughtroboticsvitcc Can you merge it?

**Review by PratikGarai — APPROVED** [review]

**Pull request merged**